### PR TITLE
Add pybind for Open3D Viewer

### PR DESCRIPTION
**Pull request by ShreyanshDarshan**
Python wrapper for the Open3D Viewer can be used like this:

```python
import open3d.app as app
app.main()
```
or by running 
```python
o3d.visualization.app.run_viewer(args)
```
The second way of running viewer is not meant to be used by the user as `args` should be of the format:
```python
args = ["path/to/open3d_python_installation/app.py", "path/to/mesh_or_ptcloud.file"]
```

Mesh or pointcloud files as arguments to the python wrapper for Open3D Viewer will be made available for the user in the command line interface PR #4594.

---
This change is [Reviewable](https://reviewable.io/reviews/isl-org/open3d/4586)

### CODE REVIEW

**Comment by theNded** [comment]
I guess it's better to have all the file names underscored, meaning `vis_gui` as done in a previous PR.

File: cpp/open3d/visualization/app/Viewer.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+namespace {
+static const std::string gUsage = "Usage: Open3DViewer [meshfile|pointcloud]";
+}  // namespace
+
+namespace open3d {
+namespace visualization {
+namespace app {
+
+int RunViewer(int argc, const char *argv[]) {
```
Comment (yxlao) [inline cpp/open3d/visualization/app/Viewer.cpp:49]: `void`

File: cpp/open3d/visualization/app/Viewer.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+// ----------------------------------------------------------------------------
+
+#include <memory>
+
+namespace open3d {
```
Comment (yxlao) [inline cpp/open3d/visualization/app/Viewer.h:29]: Remove forward declaration.

File: cpp/open3d/visualization/app/Viewer.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+namespace geometry {
+class Geometry;
+}
+namespace visualizer {
+class GuiVisualizer;
+}
+}  // namespace open3d
+
+#define WIDTH 1280
```
Comment (yxlao) [inline cpp/open3d/visualization/app/Viewer.h:38]: - In general, avoid constants in headers.
- Use `constexpr int`: 1) you can put in a namespace, 2) typed.

File: cpp/pybind/visualization/app/viewer.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+#include "pybind/docstring.h"
+
+namespace open3d {
+namespace visualization {
+namespace app {
+
+static void pybind_app_functions(py::module &m) {
+    m.def(
+            "run_viewer",
+            [](const std::vector<std::string> args) {
```
Comment (yxlao) [inline cpp/pybind/visualization/app/viewer.cpp:39]: `&`

File: cpp/pybind/visualization/app/viewer.cpp [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+                const char **argv = new const char *[args.size()];
+                for (uint it = 0; it < args.size(); it++) {
```
Comment (yxlao) [inline cpp/pybind/visualization/app/viewer.cpp:41]: `size_t`

File: cpp/pybind/visualization/app/viewer.cpp [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                    argv[it] = args[it].c_str();
+                }
+                RunViewer(args.size(), argv);
+                delete[] argv;
+                return 0;
```
Comment (yxlao) [inline cpp/pybind/visualization/app/viewer.cpp:46]: remove `return`

File: python/tools/app.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import sys
+
+
+def main():
+    args = [os.path.abspath(__file__)]
+    if len(sys.argv) > 1:
+        args.append(sys.argv[1])
+    o3d.visualization.app.run_viewer(args)
+
+    o3d.utility.reset_print_function()
```
Comment (yxlao) [inline python/tools/app.py:38]: remove

File: examples/python/gui/vis-gui.py [hunk 8]
```diff
@@ -722,30 +722,38 @@ def _on_about_ok(self):
     def load(self, path):
         self._scene.scene.clear_geometry()
 
-        geometry = None
+        geometries = []
```
Comment (yxlao) [inline examples/python/gui/vis-gui.py:725]: Move the change to a different PR.

File: cpp/open3d/visualization/app/Viewer.cpp [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    auto &app = gui::Application::GetInstance();
+    app.Initialize(argc, argv);
+
+    auto vis = std::make_shared<GuiVisualizer>("Open3D", WIDTH, HEIGHT);
+    bool is_path_valid = (path && path[0] != '\0');
+    if (is_path_valid) {
+        vis->LoadGeometry(path);
+    }
+    gui::Application::GetInstance().AddWindow(vis);
+    // when Run() ends, Filament will be stopped, so we can't be holding on
```
Comment (yxlao) [inline cpp/open3d/visualization/app/Viewer.cpp:71]: Nit: When

**Review by yxlao — CHANGES_REQUESTED** [review]

File: cpp/open3d/visualization/app/Viewer.cpp [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+#include "open3d/visualization/gui/Application.h"
+#include "open3d/visualization/gui/Native.h"
+#include "open3d/visualization/visualizer/GuiVisualizer.h"
+
+using namespace open3d;
+using namespace open3d::geometry;
+using namespace open3d::visualization;
+
+namespace {
+static const std::string gUsage = "Usage: Open3DViewer [meshfile|pointcloud]";
```
Comment (yxlao) [inline cpp/open3d/visualization/app/Viewer.cpp:42]: 1. Put inside `open3d::visualization::app`. I think it can be placed together with:
    ```cpp
    const int width = 1280;
    const int height = 960;
    ```
2. Rename `gUsage` to `snake_case`.

File: cpp/open3d/visualization/app/Viewer.cpp [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+    auto &app = gui::Application::GetInstance();
+    app.Initialize(argc, argv);
+
+    const int width = 1280;
+    const int height = 960;
+    auto vis = std::make_shared<open3d::visualization::GuiVisualizer>(
+            "Open3D", width, height);
+    bool is_path_valid = (path && path[0] != '\0');
+    if (is_path_valid) {
```
Comment (yxlao) [inline cpp/open3d/visualization/app/Viewer.cpp:70]: nit: `if (path && path[0] != '\0')`, since `is_path_valid` is only used once.

File: python/tools/app.py [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+def main():
+    args = [os.path.abspath(__file__)]
+    if len(sys.argv) > 1:
+        args.append(sys.argv[1])
+    o3d.visualization.app.run_viewer(args)
+
+
+if __name__ == "__main__":
+    sys.exit(main())
```
Comment (yxlao) [inline python/tools/app.py:40]: `main()` does not have return code, we can simply do `main()`

File: cpp/open3d/visualization/app/Viewer.h [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+// IN THE SOFTWARE.
+// ----------------------------------------------------------------------------
+
+#include <memory>
+
+namespace open3d {
+namespace visualization {
+namespace app {
+
+void RunViewer(int argc, const char *argv[]);
```
Comment (yxlao) [inline cpp/open3d/visualization/app/Viewer.h:33]: Add docstring for this function.

**Review by yxlao — CHANGES_REQUESTED** [review]

File: cpp/open3d/utility/Logging.h [hunk 14]
```diff
@@ -147,6 +147,9 @@ class Logger {
     /// Reset the print function to the default one (print to console).
     void ResetPrintFunction();
 
+    /// Get the print function used by the Logger.
+    std::function<void(const std::string &)> GetPrintFunction();
```
Comment (yxlao) [inline cpp/open3d/utility/Logging.h:151]: can this be `const`?

**Review by yxlao — APPROVED** [review]
LGTM with one more minor comment.
